crates/cext/src/transpiler/target.rs

```diff
@@ -0,0 +1,777 @@
+///
+///     QkTarget *copied = qk_target_copy(target);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_copy(target: *mut Target) -> *mut Target {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let target = unsafe { const_ptr_as_ref(target) };
+
+    Box::into_raw(target.clone().into())
+}
+
+/// @ingroup QkTarget
+/// Free the ``Target``.
+///
+/// @param target A pointer to the ``Target`` to free.
+///
+/// # Example
+///     
+///     QkTarget *target = qk_target_new(5);
+///     qk_target_free(target);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``target`` is not a valid, non-null pointer to a ``QkTarget``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_target_free(target: *mut Target) {
+    if !target.is_null() {
+        if !target.is_aligned() {
+            panic!("Attempted to free a non-aligned pointer.")
+        }
+
+        // SAFETY: We have verified the pointer is non-null and aligned, so it should be
+        // readable by Box.
+        unsafe {
+            let _ = Box::from_raw(target);
+        }
+    }
+}
+
+/// Represents the mapping between qargs and ``InstructionProperties``
+#[derive(Debug)]
+pub struct TargetEntry {
+    operation: StandardGate,
+    params: Option<SmallVec<[Param; 3]>>,
+    map: IndexMap<Qargs, Option<InstructionProperties>, ahash::RandomState>,
+}
+
+impl TargetEntry {
+    pub fn new(operation: StandardGate) -> Self {
+        Self {
+            operation,
+            params: None,
+            map: Default::default(),
+        }
+    }
+
+    pub fn new_fixed(operation: StandardGate, params: SmallVec<[Param; 3]>) -> Self {
+        Self {
+            operation,
+            params: Some(params),
+            map: Default::default(),
+        }
+    }
+}
+
+/// @ingroup QkTargetEntry
+/// Creates an entry to the ``QkTarget`` based on a ``QkGate`` instance with
+/// no parameters.
+///
+/// @param operation The Standard Gate representing this entry in the target.
+///
+/// @return The ``QkTargetEntry`` structure.
+///
+/// # Example
+///
+///     QkTargetEntry *entry = qk_target_entry_new(QkGate_H);
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEntry {
+    // Fast fail if the instruction is expecting parameters.
+    if operation.num_params() != 0 {
+        panic!("Tried to create an non-parametric entry with a parametric gate.")
+    }
+    Box::into_raw(Box::new(TargetEntry::new(operation)))
+}
+
+/// @ingroup QkTargetEntry
+/// Creates an object that will serve as a mapping between an instruction's
+/// qargs and instruction properties.
+///
+/// @param operation The Standard Gate representing this entry in the target.
+/// @param params The list of fixed parameters for this gate.
```

Comment:
This sounds like it would be a list of parameters it is calibrated for, maybe something like this? Might also be good to note the length here, even though it is repeated in the safety section below
```suggestion
/// @param params A pointer to the parameters that the instruction is calibrated for. 
///     The number of parameters must be equal to ``qk_gate_num_params(operation)``.
```